feat: new Motoko base library announcement blog post (#5502)
dfinity/portal: blog/features/new-motoko-base.md

---
title: "Try the new Motoko base library: feedback requested!"

> description: The new base library makes it easier for both humans and AI to read and write Motoko canisters.

Reviewer: ```suggestion
description: The new Motoko base library makes it easier for both humans and AI to read and write Motoko canisters.
```

tags: [Motoko, New features, Technology]
image: /img/blog/new-motoko-base.jpg
---

# Try the new Motoko base library: feedback requested!

[![New Motoko base library](/img/blog/new-motoko-base.jpg)](https://github.com/dfinity/new-motoko-base)

* [Developer forum topic](https://forum.dfinity.org/t/motoko-base-library-changes/39766)
* [GitHub repository](https://github.com/dfinity/new-motoko-base)
* [ICP Ninja demo project](https://icp.ninja/s/kwKkw)

> The Languages team is excited to announce a major overhaul of the Motoko base library! Our goal is to improve the consistency and usability of Motoko’s standard library, making it easier for both humans and AI to read and write Motoko canisters.

Reviewer: ```suggestion
The Motoko team is excited to announce a major overhaul of the Motoko base library! Our goal is to improve the consistency and usability of Motoko’s standard library, making it easier for both humans and AI to read and write Motoko canisters.
```

Author: Is it called the Motoko team now? That at least sounds better than Languages team either way.

Here is a quick summary of the biggest features and improvements:

> * New imperative and functional data structures 
> * Simplified type conversions
> * Data structures no longer rely on hashing
> * `range()` functions for each numeric type, with an exclusive upper bound
> * `VarArray` module for more conveniently working with mutable arrays
> * `Random` module with a cleaner API and optional pseudo-random number generation
> * Many other changes for consistency, clarity, and convenience

Reviewer: ```suggestion
* New imperative and functional data structures.
* Simplified type conversions.
* Data structures no longer rely on hashing.
* `range()` functions for each numeric type, with an exclusive upper bound.
* `VarArray` module for more conveniently working with mutable arrays.
* `Random` module with a cleaner API and optional pseudo-random number generation.
* Many other changes for consistency, clarity, and convenience.
```

## Try it yourself

Last week, we released a preview of the new base library with a few different options for trying it out. One is through the [`new-base`](https://mops.one/new-base) Mops package. Add the following to your `mops.toml` config file:

```toml
new-base = "0.2.0" # Check https://mops.one/new-base for the latest version
```

Alternatively, you can directly replace the `mo:base` imports in an existing project:

```toml
base = "https://github.com/dfinity/new-motoko-base"
```

> To further simplify trying out the new base library, we created an online demo which you can use to explore the new base library using ICP Ninja:
> 
> https://icp.ninja/s/kwKkw

Reviewer: ```suggestion
To further simplify trying out the new base library, we created an online demo which you can use to explore the new base library using ICP Ninja: https://icp.ninja/s/kwKkw
```

In the future, we intend to remove the `new-base` preview and distribute these changes as the official `base`. We will do our best to ensure that packages relying on previous base library versions continue to work as expected. 

If you want to use this preview for anything important, keep in mind that it's under active development. Expect future breaking changes and the possibility of bugs or outdated documentation. Please let us know if you run into something unexpected by opening a [GitHub issue](https://github.com/dfinity/new-motoko-base/issues).

## What's changed?

> Below is a detailed overview of the most notable changes and additions:

Reviewer: ```suggestion
Below is a detailed overview of the most notable changes and additions.
```

We chose implementations with good all-round performance, deferring specialized implementations to the [Mops](https://mops.one/) package ecosystem. We also updated function names for consistency and familiarity from other languages such as JS, Python, Java, and Rust.

Below is an example of using the new imperative `List` module, derived from the [`vector`](https://mops.one/vector) Mops package (big thanks to [Timo Hanke](https://github.com/timohanke)):

```motoko no-repl
import List "mo:base/List";
import Nat "mo:base/Nat";

actor {
  stable let list = List.empty<Nat>(); // Persistent data structure
  List.add(list, 5);
  assert List.toText(list, Nat.toText) == "[5]";
}
```

The above code snippet can be rewritten as a new `persistent` actor:

```motoko no-repl
import List "mo:base/List";

persistent actor {
  let list = List.empty<Nat>(); // Persistent data structure
  ...
}

> ```

Reviewer: Would this read better in a table so the two are side by side and the reader can more directly see what bits change?

Author: We could do this, although it might take a day or two to write and verify all the examples. The original code is a lot more complex, so I was thinking we could do this as part of a separate migration guide or otherwise maybe come back to this in the future once the new base library is fully stable.

> You can also use the purely functional `List` module:

Reviewer: How does the imperative `List` module differ from the purely functional one? A description of this might be helpful, and maybe an example of when one should be used over the other.

Author: Sure, that seems worth including.

Author: Added a link to an article which goes into detail on this. Readers will likely either know the difference or need to go on a significant tangent to understand this topic. I also reworded to clarify imperative vs. functional and mutable vs. immutable.

```motoko no-repl
import PureList "mo:base/pure/List";

persistent actor {
  var list = PureList.empty<Text>(); // Persistent data structure
  list := PureList.push(list, "Hi");
  assert PureList.size(list) == 1;
  assert PureList.all<Text>(list, func(n) { n == "Hi" });
}
```

We also included an efficient [stable BTree map implementation](https://github.com/canscale/StableHeapBTreeMap) (big thanks to [Byron Becker](https://github.com/ByronBecker)):

```motoko no-repl
import PureMap "mo:base/pure/Map";
import Text "mo:base/Text";
import Array "mo:base/Array";

persistent actor {
  var map = PureMap.empty<Text, Nat>();
  map := PureMap.add(map, Text.compare, "key", 123);
  assert PureMap.size(map) == 1;
  assert Array.fromIter(PureMap.entries(map)) == [("key", 123)];
}
```

Here's the complete list of data structures in the new base library:

> * List (adapted from [`vector`](https://mops.one/vector) Mops package)
> * Map
> * Queue
> * Set 
> * Stack
> * pure/List
> * pure/Map (adapted from [`StableHeapBTreeMap`](https://mops.one/stableheapbtreemap) Mops package)
> * pure/Queue
> * pure/Set

Reviewer: ```suggestion
* `List` (adapted from [`vector`](https://mops.one/vector) Mops package)
* `Map`
* `Queue`
* `Set`
* `Stack`
* `pure/List`
* `pure/Map` (adapted from [`StableHeapBTreeMap`](https://mops.one/stableheapbtreemap) Mops package)
* `pure/Queue`
* `pure/Set`
```

### Type conversions

> It's simple to convert between mutable and purely functional data structures:

Author: I guess before now there weren't really any mutable data structures, so this is all new. There might be a better way to indicate this in the blog post.

We also fixed naming inconsistencies in functions by replacing `ArrayMut` with `VarArray` across the base library.

### Random module

We completely redesigned the `Random` module, which is now a lot simpler to use:

```motoko no-repl
import Random "mo:base/Random";

actor {
  let random = Random.crypto(); // Cryptographic random numbers from ICP runtime

  public func coinFlip() : async Bool {
    await* random.bool()
  };

  public func randomItem(items : [Text]) : async Text {
    items[await* random.natRange(0, items.size())]
  };
}
```

It's now possible to use pseudo-random number generation, adapted from the [`prng`](https://mops.one/prng) Mops package (big thanks to [Timo Hanke](https://github.com/timohanke)):

```motoko no-repl
import Random "mo:base/Random";

actor {
  let seed : Nat64 = 12345;
  let random = Random.fast(seed); // Pseudo-random number generator from a seed

  public func coinFlip() : async Bool {
    random.bool()
  };

  public func randomItem(items : [Text]) : async Text {
    items[random.natRange(0, items.size())]
  };
}
```

> It's worth mentioning that the `Random` module is likely to see further changes such as adding a way to persist the state of pseudo-random number generation.

Reviewer: ```suggestion
It's worth mentioning that the `Random` module is likely to see further changes in the future, such as adding a way to persist the state of pseudo-random number generation.
```

Author: How about:

```suggestion
It's worth mentioning that the `Random` module is likely to see more changes in the future, such as adding a way to persist the state of pseudo-random number generation.
```

## What's next?

Before replacing the current Motoko base library, we have a list of follow-up improvements:

> * Overhauling the documentation with clearer explanations and more examples
> * Improving test coverage of new functionality
> * Adding new language capabilities to simplify data structure usage

Reviewer: ```suggestion
* Overhauling the documentation with clearer explanations and more examples.
* Improving test coverage of new functionality.
* Adding new language capabilities to simplify data structure usage.
```

  };
}
```

Now, you can use pseudo-random number generation, adapted from the [`prng`](https://mops.one/prng) Mops package (big thanks to [Timo Hanke](https://github.com/timohanke)):

```motoko no-repl
import Random "mo:base/Random";

actor {
  let seed : Nat64 = 12345;
  let random = Random.fast(seed); // Pseudo-random number generator from a seed

  public func coinFlip() : async Bool {
    random.bool()
  };

  public func randomItem(items : [Text]) : async Text {
    items[random.natRange(0, items.size())]
  };
}
```

It's worth mentioning that the `Random` module is likely to see more changes in the future, such as adding a way to persist the state of pseudo-random number generation.

## What's next?

Before replacing the current Motoko base library, we have a list of follow-up improvements:

* Overhauling the documentation with clearer explanations and more examples.
* Improving test coverage of new functionality.
* Adding new language capabilities to simplify data structure usage.

## Contributions and feedback

We want to give a huge thanks to the community members who provided high-quality code contributions to the new base library repository:

> * [Timo Hanke](https://github.com/timohanke): [`vector`](https://github.com/research-ag/vector), [`prng`](https://github.com/research-ag/prng)

Reviewer: ```suggestion
* MR Research AG (A. Stepanov, T. Hanke): [`vector`](https://github.com/research-ag/vector), [`prng`](https://github.com/research-ag/prng)
```

Reviewer: Requested change from Timo
